File: assets/code/4/reader.go
package main

import (
	"bufio"
	"log"
	"os"
	"syscall"
)

func main() {
	path := "./fifo"

	_, err := os.Stat(path)

	if os.IsNotExist(err) {
		if err := syscall.Mkfifo(path, 0600); err != nil {
			log.Panicln(err)
		}
	}

	f, err := os.Open(path)
	if err != nil {
		log.Panicln(err)
	}

	defer func() {
		if err := f.Close(); err != nil {
			log.Println(err)
		}
	}()

	for {
		b, err := bufio.NewReader(f).ReadByte()
		if err != nil{
			log.Panicln(err)
		}

		log.Printf("read : %d\n", b)
	}
}

File: assets/code/4/reader.py
import os

try:
    path = "./fifo"
    os.mkfifo(path, 0o600)  # try to create a named pipe
except FileExistsError:
    print("file already exists")

with open("./fifo", "r") as f:
    while 1:
        byte = f.read(1)
        print("read : {}".format(ord(byte)))
